src/box/txn_limbo.c

```diff
@@ -202,6 +204,14 @@ void
 txn_limbo_abort(struct txn_limbo *limbo, struct txn_limbo_entry *entry)
 {
 	entry->is_rollback = true;
+	if (entry == limbo->confirmed_entry) {
+		limbo->confirmed_entry = rlist_prev_entry(
+			limbo->confirmed_entry, in_queue);
+		limbo->confirmed_entry_is_valid = false;
+	} else if (entry == rlist_next_entry(limbo->confirmed_entry,
+					     in_queue)) {
+		limbo->confirmed_entry_is_valid = false;
+	}
```

Comment:
I once again read all the places where I wrote about the invariant.
> `confirmed_entry_is_valid == true` only when the next transaction after `confirmed_entry` is synchronous and has lsn (not -1) and `ack_count` stores the correct value.
If at any point this is no longer true, `confirmed_entry_id_valid` is immediately set to `false`. This is processed in all places where the limbo is modified in any way.

> But we have an invariant `confirmed_entry_is_valid == true` <=> the next transaction after `confirmed_entry` is synchronous and has lsn (not -1) and `ack_count` stores the correct value. Here it breaks down.


By calling `txn_limbo_abort` we do not know whether `confirmed_entry_is_valid` is `true`. And we don't check this in any way. And adding this check does not make sense, because it will not simplify anything. 
But what is really true is that if `entry == limbo->confirmed_entry` then `limbo->confirmed_entry_is_valid == false`, so the following patch can be applied:
```diff
void
txn_limbo_abort(struct txn_limbo *limbo, struct txn_limbo_entry *entry)
{
	entry->is_rollback = true;
	if (entry == limbo->confirmed_entry) {
		limbo->confirmed_entry = rlist_prev_entry(
			limbo->confirmed_entry, in_queue);
-		limbo->confirmed_entry_is_valid = false;
+		assert(!limbo->confirmed_entry_is_valid);
	} else if (entry == rlist_next_entry(limbo->confirmed_entry,
					     in_queue)) {
		limbo->confirmed_entry_is_valid = false;
	}
	txn_limbo_pop(limbo, entry);
}
```
Good point, thanks!